File: [0047][Medium][Permutations_II][backtracking]/Permutations_II.py
# leetcode time     cost : 48 ms
# leetcode memory   cost : 13.9 MB
# Time  Complexity: O(n*n!)
# Space Complexity: O(n*n!)
# solution 1,backtracking with swapping elements
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def backtrack(current_len = 0):
            if current_len == n:  
                res.append(nums[:])
                return
            backtrack(current_len + 1)
            for i in range(0, current_len):
                if(nums[i] == nums[current_len]):
                    return
                nums[current_len], nums[i] = nums[i], nums[current_len]
                backtrack(current_len + 1)
                nums[current_len], nums[i] = nums[i], nums[current_len]
        
        nums.sort()
        n = len(nums)
        res = []
        backtrack()
        return res
def main():
    inputX,expectRes = [1,1,2], [[1,1,2],[1,2,1],[2,1,1]]
    obj = Solution()
    result = obj.permute(inputX)
    try:
        assert result == expectRes
        print("passed, result is follow expect:",result)
    except AssertionError as aError:
        print('failed, result >> ', result,"<< is wrong, ","expect is : "+ expectRes)
    
if __name__ =='__main__':
    main()  

File: [0047][Medium][Permutations_II][backtracking]/Permutations_II.cpp
// -*- coding: utf-8 -*- 
// leetcode time     cost : 12 ms
// leetcode memory   cost : 9 MB
// Time  Complexity: O(N*N!)
// Space Complexity: O(N*N!)
// solution 1,backtrack with swapping elements
/*
全排列的问题在题解和评论区主要有两种思路：
1. 使用 used[i] 布尔数组记录元素是否使用过；
2. 使用交换的方式，在深度优先遍历（回溯）的过程中生成。
*/
// https://leetcode-cn.com/problems/permutations-ii/solution/shuang-bai-si-lu-chao-ji-jian-ji-by-tai-yang-tian/
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
private:
    vector<vector<int>> res;

    void generatePermutation(vector<int> &nums, int begin) {
        if (begin == nums.size()) {
            res.push_back(nums);
            return;
        }

        generatePermutation(nums, begin + 1);
        for (int i = 0; i < begin; i++) {
            if (nums[i] == nums[begin]) {
                break;
            }
            swap(nums[i], nums[begin]);
            generatePermutation(nums, begin + 1);
            swap(nums[i], nums[begin]);
        }
    }

public:
    vector<vector<int>> permuteUnique(vector<int> &nums) {
        sort(nums.begin(), nums.end());
        generatePermutation(nums, 0);
        return res;
    }
};